doc/tutorial/concepts/index.md

```diff
@@ -0,0 +1,319 @@
+}
+```
+
+The template parameter `t` is said to be *unconstrained*, in theory it can be instantiated with any type.
+But of course it won't actually compile for all types, because the function template **implicitly requires** that
+types provide a `+` operator.
+If a type is used that does not have a `+` operator, this implicitness causes the compiler to fail at the place where
+such operator is used – and not at the place the template is instantiated.
+This leads to very complex error messages for deeply nested code.
+
+*Constraints* are a way of making requirements of template arguments **explicit**.
+Constraints can be formulated ad-hoc, but this tutorial only covers *concepts*.
+The interested reader can check the [documentation](https://en.cppreference.com/w/cpp/language/constraints) to learn
+about ad-hoc definitions.
+Concepts are a set of constraints with a given name.
+Let's assume there is a concept called `Addable` that requires the existence of a `+` operator (as previously mentioned
+the syntax for defining concepts is not covered here).
+The following snippet demonstrates how we can constrain our function template, i.e. make the template immediately
+reject any types that don't satisfy the requirement:
+
+```cpp
+template <Addable t>
+t add(t const v1, t const v2)
+{
+    return v1 + v2;
+}
+
+int main()
+{
+    return add(1, 3); // instantiates add<int>()
+}
+```
+
+The only difference is that we have replaced `typename` with `Addable`.
+If you plug in a type that does not model `Addable`, you will get a message stating exactly that and not a cryptic
+template backtrace.
+
+The standard library provides a set of [predefined concepts](https://en.cppreference.com/w/cpp/concepts).
+For our example above, the std::Integral concept could have been used.
+
+## Syntax variants
+
+Depending on the complexity of your constraint statements, three different syntaxes are available to enforce
+constraints; all of the following are equivalent.
+
+(1) The "verbose syntax", especially useful when enforcing multiple constraints:
+
+```cpp
+template <typename t1, typename t2>
+    requires std::Integral<t1> && std::Integral<t2> // && MyOtherConcept<t1>
+auto add(t1 const v1, t2 const v2)
+{
+    return v1 + v2;
+}
+```
+
+(2) The "intermediate syntax":
+```cpp
+template <std::Integral t1, std::Integral t2>                   // one constraint per type
+auto add(t1 const v1, t2 const v2)
+{
+    return v1 + v2;
+}
+```
+
+(3) The "terse syntax":
+```cpp
+auto add(std::Integral auto const v1, std::Integral const v2)   // one constraint per type
+{
+    return v1 + v2;
+}
+```
+
+\attention The terse syntax in this form is not yet available in GCC7, GCC8 and GCC9.
+
+Different constraints can be applied to different template parameters and a single template parameter can be constrained
+by multiple concepts.
+Syntaxes can also be combined:
+```cpp
+template <std::Integral t1, std::Integral t2>
+    // requires MyOtherConcept<t1>
```

Comment:
why is this commented out? It is not a compiled snippet anyway. 
I would prefer a real example.

Reply:
Actually this is already an example of how we use things within the library. So maybe add a sentence like: `Within SeqAn we always use the intermediate syntax and additionally use the verbose expressions if the intermediate constraint needs to be refined further with more constraints.`